doc/arch/011-rollup-summary-tables-for-dashboards.md

```diff
@@ -0,0 +1,65 @@
+# ADR 11 Rollup summary tables for Dashboards
+
+## Context
+Earlier we went with a [materialized view approach](https://github.com/simpledotorg/simple-server/blob/master/doc/arch/008-materialized-views-for-dashboards.md) for dashboard reporting data. This approach has worked out well for the current reports but has limitations in terms of scaling and complexity.
+
+We currently have a need for a new set of dashboard reports that rely on similar sets of data, and anticipate more needs for summary data for monthly and quarterly reports. We'd like to try to find a simpler, more stable, and more scalable approach to making that data easily accessible.
+
+## Decision
+
+We will create rollup summary tables for to capture the "most recent blood pressure per patient per period". These rollup tables will be updated in real-time as blood pressures are captured, removing the need for recurring batch jobs. We will denormalize as much data as we need to to make reporting queries simple and cheap to execute, to prevent the need for any complicated SQL queries or additional data layer caching.
```

Comment:
> most recent blood pressure per patient per period

This sounds like the views that we have. 

Could we better establish some actual bottlenecks on why a daily batch job that refreshes views is problematic and a real-time job over every sync call helps mitigate those problems?

Reply:
> This sounds awfully similar to the views that we have 😉

Agreed 😄 . I think the fundamental difference is that the rollup table will be updated once for each new blood pressure that comes into the system. Once a BP is entered on May 29th for a patient, that row for that patient won't have to be updated or touched again (assuming the patient doesn't come back for more BP readings in May). 

With the materialized views, we are taking up CPU / DB time writing that May 29th BP, _all other_ BPs every time the views get refreshed. So we are doing way more work than we need to with the views vs a rollup summary approach.

> Could we better establish some actual bottlenecks on why a daily batch job that refreshes views is problematic and a real-time job over every sync call helps mitigate those problems?

I would love to, but we need better metrics and better visibility in the system first to do that. Something like grafana or prometheus, or maybe we could use custom new relic tracing.  

If its not a problem right now, I can be pretty sure that recomputing the entire blood pressure table into views will be a problem when we have tens of millions of blood pressures in the system.